### PR TITLE
Add FollowMeMusic binding

### PR DESCRIPTION
**Pull request by MattPieraggi**
New binding implementing a localisation algorithm used in a musical system:

The goal of our project is to allow to one or several users to listen to music in their own house by moving but without interacting with the system. The sound, corresponding to the personnal music or to the radio, will go on or off in each room depending on the presence of the user in that specific room thanks to our connected speakers.

To accomplish this task, we installed sensors next to the doors in order to detect users with their own personnal strap, these informations will be sent to the principal box which will treat them and will turn on or off the connected speakers concerned. Every user has his own personnal strap so he can be identified among the other users. So there can be differents musics playing in differents rooms of the house.
	
By passing from a room to another with his strap, the user activates the first sensor on one side of the door then the other one on the other side. That permits to the system to know this user passed the door in one specific way.

At any moment, each user can add a strap, a speaker or sensors in his housing environment. He only has to connect it to the network, then to synchronize it with the system by defining the characteristics of each equipment.

### CODE REVIEW

**Comment by teichsta** [comment]
Thanks for this contribution @MattPieraggi. Will start reviewing it soon. Best, Thomas E.-E.

File: bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/followmemusic/location/LocationManager.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+			
+			// Notify
+			if(callback != null) {
+				callback.success(user, newRoom, previousRoom);
+			}
+			
+		/*} catch (UserLocationException e) {
+			e.printStackTrace();
+			callback.failure(user);
+		}*/
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/followmemusic/location/LocationManager.java]: remove commented code

File: bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/followmemusic/location/DatabaseLocationManager.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+					val.put("index", index);
+					val.put("max", max);
+					
+					thresholdSensors.add(val);
+				}
+			}
+			
+			if(thresholdSensors.size() > 0)
+			{
+				logger.warn("------------- ALGO COMPUTING -----------");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/followmemusic/location/DatabaseLocationManager.java]: at least decrease to DEBUG or better remove or enhance with reasonable debug information

File: bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/followmemusic/notification/UpdateLocationNotification.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package org.followmemusic.notification;
+
+public class UpdateLocationNotification extends TCPNotification {
+	
+	private static final String targetIP = "127.0.0.1";
+	private static final int port = 4545;
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/followmemusic/notification/UpdateLocationNotification.java]: shouldn't these configuration parameters are configurable?

File: bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/openhab/binding/followmemusic/internal/FollowMeMusicBinding.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+		}
+	}
+		
+	
+	public void activate() {
+	}
+	
+	public void deactivate() {
+		// deallocate resources here that are no longer needed and 
+		// should be reset when activating this binding again
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/openhab/binding/followmemusic/internal/FollowMeMusicBinding.java]: remove generic comments

File: bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/openhab/binding/followmemusic/internal/FollowMeMusicBinding.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	/**
+	 * @{inheritDoc}
+	 */
+	@Override
+	protected void internalReceiveCommand(String itemName, Command command) {
+		// the code being executed when a command was sent on the openHAB
+		// event bus goes here. This method is only called if one of the 
+		// BindingProviders provide a binding for the given 'itemName'.
+		logger.debug("internalReceiveCommand() is called!");
+		logger.debug(itemName+" = "+command.toString());
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/openhab/binding/followmemusic/internal/FollowMeMusicBinding.java]: remove generic comment and logging

File: bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/openhab/binding/followmemusic/internal/FollowMeMusicBinding.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	/**
+	 * @{inheritDoc}
+	 */
+	@Override
+	protected void internalReceiveUpdate(String itemName, State newState) {
+		// the code being executed when a state was sent on the openHAB
+		// event bus goes here. This method is only called if one of the 
+		// BindingProviders provide a binding for the given 'itemName'.
+		logger.debug("internalReceiveCommand() is called!");
+	}
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.followmemusic/src/main/java/org/openhab/binding/followmemusic/internal/FollowMeMusicBinding.java]: remove this overridden method

**Comment by teichsta** [comment]
Hi @MattPieraggi,

here are my first comments regarding your Binding:

* format the code using the standard Eclipse Code Formatter
* add license header to ALL files (use mvn -Dyear=2015 license:format to accomplish that easily)
* add javadoc on classlevel and for each public method (using @author and @since 1.7.0 on classlevel)
* use english language to document the code
* member variables should all be declared at the top of the class (after the class keyword)
* replace all printStacktrace() with proper logging. When choosing the log level please keep in mind whether the given error is really an error with regards to the platform?
* decrease log LEVEL where reasonable (found some WARN that obviously are DEBUGs, e.g. DatabaseLocationManager.javaL194)
* what is the license of the code under "org.followmemusic.*" ?
* add your binding to bundles/binding/pom.xml
* add your binding to distribution/pom.xml
* add the given configuration parameters to openhab_default.cfg (new section for FollowMeMusic)

In general i am wondering if using a fullblown SQL database is really necessary to administer the given sensors and locations. Do you have any numbers on how the DB is utilised? Solving the follow-me problem also seem to very realtime … so there would also be no need to work on historic data, right?

Also i probably didn't get the purpose of the Binding at all. As far as i can see it "only" tracks people in rooms and knows/updates (does it) about their location. However this doesn't seem to specific to music, isn't it?

Could you probably explain a little more how it relates to music?

Best, Thomas E.-E.

**Comment by teichsta** [comment]
rescheduled for 1.8

**Comment by teichsta** [comment]
any updates @MattPieraggi?

**Comment by MattPieraggi** [comment]
Sorry, the team I've been working with and myself no longer work on this project because of lack of time.

**Pull request closed without merging**